Create XML Merge function

LucasRochaAbraao opened the pull request:
Create a function that edits a junit xml produced by pytest and updates test results in the original file.

EDEN-2543

lrochaab committed:
Create XML Merge function

Create a function that edits a junit xml produced by pytest and updates
test results in the original file.

EDEN-2543

igortg commented on conftest.py:
```diff
@@ -0,0 +1,37 @@
+
+from _pytest.reports import TestReport
+
+from src.pytest_merge_xml.merge_xml import modify_xml
```
This should be:
```suggestion
from pytest_merge_xml.merge_xml import modify_xml
```

raispbizerra commented on conftest.py:
```diff
@@ -0,0 +1,37 @@
+
+def pytest_configure(config):
+    original_xmlfile = config.getoption("--mergexml")
+    # report = XMLReport(Path('src/pytest_merge_xml/failure.xml'))
```
Left-over?

LucasRochaAbraao replied:
Yeah, I'll remove it.

raispbizerra commented on src/pytest_merge_xml/merge_xml.py:
```diff
@@ -0,0 +1,27 @@
+    ]
+
+    for testcase in testsuite:
+        testcase_name = testcase.attrib['name']  # 'test_xml_one_failed_two_passed'
```
What's this comment about?

LucasRochaAbraao replied:
I was using it as reference during development, I'll remove it as well!

nicoddemus commented on pyptoject.toml:
```diff
@@ -1,7 +1,8 @@
 [tools.isort]
```
The name of the file is wrong, should be `pyproject.toml`

LucasRochaAbraao replied:
Thanks 👍

igortg commented on src/pytest_merge_xml/merge_xml.py:
```diff
@@ -0,0 +1,27 @@
+    for testcase in testsuite:
+        testcase_name = testcase.attrib['name']  # 'test_xml_one_failed_two_passed'
+        for failure in testcase.findall("failure"):
+            print(testcase_name, results_passed)
```
Remove this `print` statement.

LucasRochaAbraao replied:
Removed the statement.

igortg commented on src/pytest_merge_xml/merge_xml.py:
```diff
@@ -0,0 +1,27 @@
+    tree = Et.parse(original_xml)
+    root = tree.getroot()
+    testsuite = root.find("testsuite")
+    failure_count = int(testsuite.attrib['failures'])
```
```suggestion
    previous_run_failure_count = int(testsuite.attrib['failures'])
```

LucasRochaAbraao replied:
Renamed the variable.

igortg commented on .gitignore:
```diff
@@ -18,3 +18,7 @@ dist/
 .tox/
 .venv/
+
+# local files
```
Do not ignore thoses.

LucasRochaAbraao replied:
Fixed it, not ignoring the failure.xml file anymore. The other one isn't needed, I was using it to keep the original unchanged so I didn't have to recreate it.

igortg commented on conftest.py:
```diff
@@ -0,0 +1,37 @@
+from src.pytest_merge_xml.merge_xml import modify_xml
+
+
+class XMLReport:
```
Add some docstring:
Add a docstring. Something like:
```suggestion
class XMLReport:
    '''Class that collect test items and update and existing XML file with the new test results'''
    
```

LucasRochaAbraao replied:
Done!

nicoddemus commented on .pre-commit-config.yaml:
```diff
@@ -9,6 +9,6 @@ repos:
-      name: isort (python)
+  hooks:
+  - id: isort
+    name: isort (python)
```
Missing `black` for formatting

LucasRochaAbraao replied:
It's now included, thanks!

nicoddemus commented on conftest.py:
```diff
@@ -0,0 +1,37 @@
+from pathlib import Path
```
If this is meant to be a plugin installed by others, you cannot use a `conftest.py` file for this, you make this a normal module and use [setuptools entry points to register it](https://docs.pytest.org/en/stable/how-to/writing_plugins.html#making-your-plugin-installable-by-others).

LucasRochaAbraao replied:
Thanks for the correction! I changed it and it's working now. I installed the package locally and tested in a fake project to make sure.

nicoddemus commented on conftest.py:
```diff
@@ -0,0 +1,37 @@
+    # pytest has a note to keep conftest.py at the root dir in order to load added options.
+    # http://pytest.org/latest/plugins.html?highlight=hooks#plugin-discovery-order-at-tool-startup
+    parser.addoption(
+        "--mergexml",
```
I find it a bit confusing that the plugin is named `pytest-merge-xml` and the flag is `--mergexml`, I suggest:

```suggestion
        "--merge-xml",
```

LucasRochaAbraao replied:
Makes sense, thanks for the suggestion! I changed it.

lrochaab committed:
Implemented suggestions to initial PR.

Fixes:
  - Removed unnecessary comments and print statements, and added
    docstrings.
  - Fixed filename typo.
  - Fixed imports
  - Added black formatting and renamed the cli flag to be more consistent.
  - moved conftest.py into a plugin module, in order to make the project
    installable as a plugin in other projects.

EDEN-2543

lrochaab committed:
Implemented suggestions to initial PR.

Fixes:
  - Removed unnecessary comments and print statements, and added
    docstrings.
  - Fixed filename typo.
  - Fixed imports
  - Added black formatting and renamed the cli flag to be more consistent.
  - moved conftest.py into a plugin module, in order to make the project
    installable as a plugin in other projects.

EDEN-2543

nicoddemus commented on .pre-commit-config.yaml:
```diff
@@ -9,6 +9,11 @@ repos:
-    - id: isort
-      name: isort (python)
+  hooks:
+  - id: isort
```
@igortg do you feel strongly about `isort`? I personally have been using [reorder-python-imports](https://pypi.org/project/reorder-python-imports/) in all my repos.

igortg replied:
Better... @LucasRochaAbraao will do the change.

LucasRochaAbraao replied:
Changed it.

nicoddemus commented on setup.cfg:
```diff
@@ -3,3 +3,4 @@ universal = 1
 
 [flake8]
 exclude = docs
+max-line-length = 100
```
Add a `mypy` session here, using this one as basis: https://github.com/ESSS/pytest-regressions/blob/master/mypy.ini

Make sure to include `disallow_untyped_defs = True` too, to ensure all functions/methods are all type-annotated.

LucasRochaAbraao replied:
I'll look into this in a later issue!

nicoddemus commented on src/pytest_merge_xml/plugin.py:
```diff
@@ -0,0 +1,42 @@
+from pytest_merge_xml.merge_xml import modify_xml
+
+
+class XMLReport:
```
```suggestion
class XMLReportPlugin:
```

Also XMLReport does not really conveys what the class means, how about `XMLReportMergerPlugin`, or something like that?

LucasRochaAbraao replied:
In terms of conveying what the class means, I'm leaning towards 'XMLReportMergerPlugin'. 'XMLReportPlugin' also feels a bit generic, and this class actually modifies the resulting xml. Thanks for the suggestion!

nicoddemus commented on test/test_pytest_merge_xml.py:
```diff
@@ -0,0 +1,156 @@
+    file_regression.check(obtained_file.read_text(), extension=".xml")
+
+
+def test_xml_three_failed_zero_passed(
```
Those tests are all valid, but I suggest at least one integration test using the `pytester` fixture.

LucasRochaAbraao replied:
I'll look into this in a later issue! Thanks.

nicoddemus reviewed (changes requested):
I think we are almost there, please take a look at my comments. :+1:

igortg commented:
Right now I'd like to redirect all the effort to validate that this plugin would really work on our environment (being called within `inv test-py`).

I created [EDEN-2553](https://eden.esss.co/jira/browse/EDEN-2553) and [EDEN-2554](https://eden.esss.co/jira/browse/EDEN-2554) so we can tackle @nicoddemus suggestions as soon as we make make this validation.

lrochaab committed:
Merge branch 'fb-EDEN-2543-initial-files' of github.com:ESSS/pytest-merge-xml into fb-EDEN-2543-initial-files

lrochaab committed:
Merge branch 'fb-EDEN-2543-initial-files' of github.com:ESSS/pytest-merge-xml into fb-EDEN-2543-initial-files

lrochaab committed:
Merge branch 'fb-EDEN-2543-initial-files' of github.com:ESSS/pytest-merge-xml into fb-EDEN-2543-initial-files
